From be7e3a766365493a068b5a5df9f7ac16d803f55b Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Wed, 16 Jul 2014 20:42:25 -0400
Subject: [md] dm-cache-metadata: do not allow the data block size to change

Message-id: <1405543364-20901-22-git-send-email-snitzer@redhat.com>
Patchwork-id: 85903
O-Subject: [RHEL7.1 PATCH 21/40] dm cache metadata: do not allow the data block size to change
Bugzilla: 1117872
RH-Acked-by: Mikulas Patocka <mpatocka@redhat.com>
RH-Acked-by: Heinz Mauelshagen <heinzm@redhat.com>
RH-Acked-by: Joe Thornber <thornber@redhat.com>

BZ: 1117872

Upstream commit 048e5a07f282c57815b3901d4a68a77fa131ce0a
Author: Mike Snitzer <snitzer@redhat.com>
Date:   Mon Jul 14 16:59:39 2014 -0400

    dm cache metadata: do not allow the data block size to change

    The block size for the dm-cache's data device must remained fixed for
    the life of the cache.  Disallow any attempt to change the cache's data
    block size.

    Signed-off-by: Mike Snitzer <snitzer@redhat.com>
    Acked-by: Joe Thornber <ejt@redhat.com>
    Cc: stable@vger.kernel.org

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/dm-cache-metadata.c b/drivers/md/dm-cache-metadata.c
index 4ead4ba..d2899e7 100644
--- a/drivers/md/dm-cache-metadata.c
+++ b/drivers/md/dm-cache-metadata.c
@@ -425,6 +425,15 @@ static int __open_metadata(struct dm_cache_metadata *cmd)
 
  disk_super = dm_block_data(sblock);
 
+ /* Verify the data block size hasn't changed */
+ if (le32_to_cpu(disk_super->data_block_size) != cmd->data_block_size) {
+  DMERR("changing the data block size (from %u to %llu) is not supported",
+        le32_to_cpu(disk_super->data_block_size),
+        (unsigned long long)cmd->data_block_size);
+  r = -EINVAL;
+  goto bad;
+ }
+
  r = __check_incompat_features(disk_super, cmd);
  if (r < 0)
   goto bad;
-- 
1.7.1